File: flutter/nepali_date_picker_kmp/ios/nepali_date_picker_kmp/Sources/nepali_date_picker_kmp/EngineApiImpl.swift
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

import Foundation
import nepali_date_picker

/// The engine host: every call is a synchronous delegation to the shared
/// `NepaliDateConverter` and its companions inside the Kotlin framework.
///
/// Kotlin/Native exceptions cannot cross into Swift, so the Dart layer
/// validates inputs before calling; everything reaching here is in range.
final class EngineApiImpl: EngineApi {

  private let converter = NepaliDateConverter.shared
  private let defaults = NepaliCalendarDefaults.shared

  func getBsYearRange() throws -> YearRangeDto { defaults.NepaliYearRange.toDto }

  func getAdYearRange() throws -> YearRangeDto { defaults.EnglishYearRange.toDto }

  func getAdYearRangeForBsYears(first: Int64, last: Int64) throws -> YearRangeDto {
    defaults.gregorianYearRangeFor(
      nepaliYearRange: KotlinIntRange(start: Int32(first), endInclusive: Int32(last))
    ).toDto
  }

  func getTodayBs() throws -> CalendarDto { converter.todayNepaliCalendar.toDto }

  func getTodayAd() throws -> CalendarDto { converter.todayEnglishCalendar.toDto }

  func getCurrentTime() throws -> TimeDto { converter.currentTime.toDto }

  func convertAdToBs(year: Int64, month: Int64, dayOfMonth: Int64) throws -> CalendarDto {
    converter.convertEnglishToNepali(
      englishYYYY: Int32(year), englishMM: Int32(month), englishDD: Int32(dayOfMonth)
    ).toDto
  }

  func convertBsToAd(year: Int64, month: Int64, dayOfMonth: Int64) throws -> CalendarDto {
    converter.convertNepaliToEnglish(
      nepaliYYYY: Int32(year), nepaliMM: Int32(month), nepaliDD: Int32(dayOfMonth)
    ).toDto
  }

  func getBsCalendar(year: Int64, month: Int64, dayOfMonth: Int64) throws -> CalendarDto {
    converter.getNepaliCalendar(
      nepaliYYYY: Int32(year), nepaliMM: Int32(month), nepaliDD: Int32(dayOfMonth)
    ).toDto
  }

  func getAdCalendar(year: Int64, month: Int64, dayOfMonth: Int64) throws -> CalendarDto {
    converter.getEnglishCalendar(
      englishYYYY: Int32(year), englishMM: Int32(month), englishDD: Int32(dayOfMonth)
    ).toDto
  }

  func isAdDateConvertible(year: Int64, month: Int64, dayOfMonth: Int64) throws -> Bool {
    converter.isEnglishDateConvertible(
      englishYYYY: Int32(year), englishMM: Int32(month), englishDD: Int32(dayOfMonth)
    )
  }

  func getBsMonth(year: Int64, month: Int64) throws -> MonthInfoDto {
    converter.getNepaliMonthCalendar(nepaliYear: Int32(year), nepaliMonth: Int32(month)).toDto
  }

  func getAdMonth(year: Int64, month: Int64) throws -> MonthInfoDto {
    converter.getEnglishMonthCalendar(englishYear: Int32(year), englishMonth: Int32(month)).toDto
  }

  func getBsCalendarsInAdMonth(year: Int64, month: Int64) throws -> [CalendarDto?] {
    converter.getNepaliCalendarsInEnglishMonthByDay(
      englishYear: Int32(year), englishMonth: Int32(month)
    ).map { $0.nepaliCalendar?.toDto }
  }

  func getAdCalendarsInBsMonth(year: Int64, month: Int64) throws -> [CalendarDto] {
    converter.getEnglishCalendarsInNepaliMonth(
      nepaliYear: Int32(year), nepaliMonth: Int32(month)
    ).map { $0.toDto }
  }

  func getTotalDaysInBsMonth(year: Int64, month: Int64) throws -> Int64 {
    Int64(converter.getTotalDaysInNepaliMonth(year: Int32(year), month: Int32(month)))
  }

  func getTotalDaysInAdMonth(year: Int64, month: Int64) throws -> Int64 {
    Int64(converter.getTotalDaysInEnglishMonth(year: Int32(year), month: Int32(month)))
  }

  func addDaysToBsDate(year: Int64, month: Int64, dayOfMonth: Int64, days: Int64) throws
    -> CalendarDto
  {
    converter.getNepaliCalendarAfterAdditionOrSubtraction(
      year: Int32(year),
      month: Int32(month),
      dayOfMonth: Int32(dayOfMonth),
      daysToAdjust: Int32(days)
    ).toDto
  }

  func getBsDaysBetween(start: DateDto, end: DateDto) throws -> Int64 {
    Int64(converter.getNepaliDaysInBetween(startDate: start.toCore, endDate: end.toCore))
  }

  func getAdDaysBetween(start: DateDto, end: DateDto) throws -> Int64 {
    Int64(converter.getEnglishDaysInBetween(startDate: start.toCore, endDate: end.toCore))
  }

  func compareBsDates(from: DateDto, to: DateDto) throws -> Int64 {
    Int64(
      converter.compareDates(
        simpleDate: from.toCore,
        year: Int32(to.year),
        month: Int32(to.month),
        dayOfMonth: Int32(to.dayOfMonth)
      ))
  }

  func getWeekdayName(dayOfWeek: Int64, format: NameFormatDto, language: LangDto) throws
    -> String
  {
    converter.getWeekdayName(
      dayOfWeek: Int32(dayOfWeek), format: format.toCore, language: language.toCore)
  }

  func getBsMonthName(month: Int64, format: NameFormatDto, language: LangDto) throws -> String {
    converter.getMonthName(month: Int32(month), format: format.toCore, language: language.toCore)
  }

  func getAdMonthName(month: Int64, format: NameFormatDto, language: LangDto) throws -> String {
    converter.getEnglishMonthName(
      month: Int32(month), format: format.toCore, language: language.toCore)
  }

  func formatBsDate(calendar: CalendarDto, locale: LocaleDto) throws -> String {
    converter.formatNepaliDate(customCalendar: calendar.toCore, locale: locale.toCore)
  }

  func formatAdDate(calendar: CalendarDto, locale: LocaleDto) throws -> String {
    converter.formatEnglishDate(customCalendar: calendar.toCore, locale: locale.toCore)
  }

  func formatBsDateByPattern(pattern: String, date: DateDto, language: LangDto) throws -> String {
    converter.formatNepaliDateByUnicodePattern(
      unicodePattern: pattern,
      calendar: converter.getNepaliCalendar(
        nepaliYYYY: Int32(date.year), nepaliMM: Int32(date.month), nepaliDD: Int32(date.dayOfMonth)
      ),
      language: language.toCore
    )
  }

  func formatAdDateByPattern(pattern: String, date: DateDto, language: LangDto) throws -> String {
    converter.formatEnglishDateByUnicodePattern(
      unicodePattern: pattern,
      calendar: converter.getEnglishCalendar(
        englishYYYY: Int32(date.year), englishMM: Int32(date.month), englishDD: Int32(date.dayOfMonth)
      ),
      language: language.toCore
    )
  }

  func formatTimeByPattern(pattern: String, time: TimeDto, language: LangDto) throws -> String {
    converter.formatTimeByUnicodePattern(
      unicodePattern: pattern, time: time.toCore, language: language.toCore)
  }

  func formatBsDateTimeByPattern(
    pattern: String, date: DateDto, time: TimeDto?, language: LangDto
  ) throws -> String {
    converter.formatNepaliDateTimeByUnicodePattern(
      unicodePattern: pattern,
      calendar: converter.getNepaliCalendar(
        nepaliYYYY: Int32(date.year), nepaliMM: Int32(date.month), nepaliDD: Int32(date.dayOfMonth)
      ),
      time: time?.toCore,
      language: language.toCore
    )
  }

  func formatAdDateTimeByPattern(
    pattern: String, date: DateDto, time: TimeDto?, language: LangDto
  ) throws -> String {
    converter.formatEnglishDateTimeByUnicodePattern(
      unicodePattern: pattern,
      calendar: converter.getEnglishCalendar(
        englishYYYY: Int32(date.year), englishMM: Int32(date.month), englishDD: Int32(date.dayOfMonth)
      ),
      time: time?.toCore,
      language: language.toCore
    )
  }

  func formatTimeEnglish(time: TimeDto, use12HourFormat: Bool) throws -> String {
    converter.getFormattedTimeInEnglish(simpleTime: time.toCore, use12HourFormat: use12HourFormat)
  }

  func formatTimeNepali(time: TimeDto, use12HourFormat: Bool) throws -> String {
    converter.getFormattedTimeInNepali(simpleTime: time.toCore, use12HourFormat: use12HourFormat)
  }

  func bsDateTimeToIso(date: DateDto, time: TimeDto) throws -> String {
    converter.formatNepaliDateTimeToIsoFormat(nepaliDate: date.toCore, time: time.toCore)
  }

  func adDateTimeToIso(date: DateDto, time: TimeDto) throws -> String {
    converter.formatEnglishDateNepaliTimeToIsoFormat(englishDate: date.toCore, time: time.toCore)
  }

  func bsDateTimeFromIso(isoDateTime: String) throws -> DateTimeDto {
    converter.getNepaliDateTimeFromIsoFormat(isoDateTime: isoDateTime).toDto
  }

  func adDateTimeFromIso(isoDateTime: String) throws -> DateTimeDto {
    converter.getEnglishDateNepaliTimeFromIsoFormat(isoDateTime: isoDateTime).toDto
  }

  func localizeDigits(text: String, script: DigitScriptDto) throws -> String {
    converter.localizeDigits(text, script: script.toCore)
  }

  func toLatinDigits(text: String) throws -> String {
    converter.toLatinDigits(text)
  }

  func wireFormatDate(date: DateDto, pattern: DatePatternDto, script: DigitScriptDto) throws
    -> String
  {
    NepaliDateFormatter.shared.format(
      date: date.toCore, pattern: pattern.toCore, script: script.toCore)
  }

  func wireParseDate(input: String, pattern: DatePatternDto) throws -> DateDto? {
    NepaliDateFormatter.shared.parse(input: input, pattern: pattern.toCore)?.toDateDto
  }

  func wireFormatTime(time: TimeDto) throws -> String {
    NepaliTimeFormatter.shared.format(time: time.toCore)
  }

  func wireParseTime(input: String) throws -> TimeDto? {
    NepaliTimeFormatter.shared.parse(input: input)?.toDto
  }

  func statusOf(policy: PolicyDto, date: DateDto) throws -> DayStatusDto {
    policy.toCore.statusOf(date: date.toCore).toDto
  }

  func monthStatus(policy: PolicyDto, year: Int64, month: Int64) throws -> [DayStatusDto] {
    policy.toCore.monthStatus(year: Int32(year), month: Int32(month)).map { $0.toDto }
  }

  func eventsOn(policy: PolicyDto, date: DateDto) throws -> [EventDto] {
    policy.toCore.eventsOn(date: date.toCore).map { $0.toDto }
  }

  func eventsIn(policy: PolicyDto, year: Int64, month: Int64) throws -> [EventDto] {
    policy.toCore.eventsIn(year: Int32(year), month: Int32(month)).map { $0.toDto }
  }

  func isNonWorkingDay(policy: PolicyDto, date: DateDto) throws -> Bool {
    policy.toCore.isNonWorkingDay(date: date.toCore)
  }

  func workingDaysBetween(policy: PolicyDto, start: DateDto, end: DateDto) throws -> Int64 {
    Int64(
      converter.workingDaysBetween(start: start.toCore, end: end.toCore, policy: policy.toCore))
  }

  func nextWorkingDay(policy: PolicyDto, from: DateDto) throws -> CalendarDto {
    let day = converter.nextWorkingDay(from: from.toCore, policy: policy.toCore)
    return converter.getNepaliCalendar(
      nepaliYYYY: day.year, nepaliMM: day.month, nepaliDD: day.dayOfMonth
    ).toDto
  }

  func addWorkingDays(policy: PolicyDto, from: DateDto, days: Int64) throws -> CalendarDto {
    let day = converter.addWorkingDays(
      from: from.toCore, days: Int32(days), policy: policy.toCore)
    return converter.getNepaliCalendar(
      nepaliYYYY: day.year, nepaliMM: day.month, nepaliDD: day.dayOfMonth
    ).toDto
  }

  func eventSpanningDays(event: EventDto, days: Int64) throws -> [EventDto] {
    event.toCore.spanningDays(days: Int32(days)).map { $0.toDto }
  }

  func eventSpanningThrough(event: EventDto, end: DateDto) throws -> [EventDto] {
    event.toCore.spanningThrough(end: end.toCore).map { $0.toDto }
  }
}

File: flutter/nepali_date_picker_kmp/android/src/main/kotlin/dev/shivathapaa/nepali_date_picker_kmp/EngineApiImpl.kt
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

package dev.shivathapaa.nepali_date_picker_kmp

import dev.shivathapaa.nepalidatepickerkmp.calendar_model.NepaliCalendarDefaults
import dev.shivathapaa.nepalidatepickerkmp.calendar_model.NepaliDateConverter
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliDateFormatter
import dev.shivathapaa.nepalidatepickerkmp.data.NepaliTimeFormatter
import dev.shivathapaa.nepalidatepickerkmp.event.addWorkingDays
import dev.shivathapaa.nepalidatepickerkmp.event.nextWorkingDay
import dev.shivathapaa.nepalidatepickerkmp.event.spanningDays
import dev.shivathapaa.nepalidatepickerkmp.event.spanningThrough
import dev.shivathapaa.nepalidatepickerkmp.event.workingDaysBetween

/**
 * The engine host: every call is a synchronous delegation to the shared
 * `NepaliDateConverter` and its companions. Kotlin exceptions surface to
 * Dart as platform errors through pigeon's wrapper.
 */
internal class EngineApiImpl : EngineApi {

    override fun getBsYearRange(): YearRangeDto =
        NepaliCalendarDefaults.NepaliYearRange.toDto()

    override fun getAdYearRange(): YearRangeDto =
        NepaliCalendarDefaults.EnglishYearRange.toDto()

    override fun getAdYearRangeForBsYears(first: Long, last: Long): YearRangeDto =
        NepaliCalendarDefaults.gregorianYearRangeFor(first.toInt()..last.toInt()).toDto()

    override fun getTodayBs(): CalendarDto =
        NepaliDateConverter.todayNepaliCalendar.toDto()

    override fun getTodayAd(): CalendarDto =
        NepaliDateConverter.todayEnglishCalendar.toDto()

    override fun getCurrentTime(): TimeDto = NepaliDateConverter.currentTime.toDto()

    override fun convertAdToBs(year: Long, month: Long, dayOfMonth: Long): CalendarDto =
        NepaliDateConverter.convertEnglishToNepali(
            year.toInt(), month.toInt(), dayOfMonth.toInt()
        ).toDto()

    override fun convertBsToAd(year: Long, month: Long, dayOfMonth: Long): CalendarDto =
        NepaliDateConverter.convertNepaliToEnglish(
            year.toInt(), month.toInt(), dayOfMonth.toInt()
        ).toDto()

    override fun getBsCalendar(year: Long, month: Long, dayOfMonth: Long): CalendarDto =
        NepaliDateConverter.getNepaliCalendar(
            year.toInt(), month.toInt(), dayOfMonth.toInt()
        ).toDto()

    override fun getAdCalendar(year: Long, month: Long, dayOfMonth: Long): CalendarDto =
        NepaliDateConverter.getEnglishCalendar(
            year.toInt(), month.toInt(), dayOfMonth.toInt()
        ).toDto()

    override fun isAdDateConvertible(year: Long, month: Long, dayOfMonth: Long): Boolean =
        NepaliDateConverter.isEnglishDateConvertible(
            year.toInt(), month.toInt(), dayOfMonth.toInt()
        )

    override fun getBsMonth(year: Long, month: Long): MonthInfoDto =
        NepaliDateConverter.getNepaliMonthCalendar(year.toInt(), month.toInt()).toDto()

    override fun getAdMonth(year: Long, month: Long): MonthInfoDto =
        NepaliDateConverter.getEnglishMonthCalendar(year.toInt(), month.toInt()).toDto()

    override fun getBsCalendarsInAdMonth(year: Long, month: Long): List<CalendarDto?> =
        NepaliDateConverter.getNepaliCalendarsInEnglishMonthByDay(year.toInt(), month.toInt())
            .map { it.nepaliCalendar?.toDto() }

    override fun getAdCalendarsInBsMonth(year: Long, month: Long): List<CalendarDto> =
        NepaliDateConverter.getEnglishCalendarsInNepaliMonth(year.toInt(), month.toInt())
            .map { it.toDto() }

    override fun getTotalDaysInBsMonth(year: Long, month: Long): Long =
        NepaliDateConverter.getTotalDaysInNepaliMonth(year.toInt(), month.toInt()).toLong()

    override fun getTotalDaysInAdMonth(year: Long, month: Long): Long =
        NepaliDateConverter.getTotalDaysInEnglishMonth(year.toInt(), month.toInt()).toLong()

    override fun addDaysToBsDate(
        year: Long,
        month: Long,
        dayOfMonth: Long,
        days: Long
    ): CalendarDto = NepaliDateConverter.getNepaliCalendarAfterAdditionOrSubtraction(
        year.toInt(), month.toInt(), dayOfMonth.toInt(), days.toInt()
    ).toDto()

    override fun getBsDaysBetween(start: DateDto, end: DateDto): Long =
        NepaliDateConverter.getNepaliDaysInBetween(start.toCore(), end.toCore()).toLong()

    override fun getAdDaysBetween(start: DateDto, end: DateDto): Long =
        NepaliDateConverter.getEnglishDaysInBetween(start.toCore(), end.toCore()).toLong()

    override fun compareBsDates(from: DateDto, to: DateDto): Long =
        NepaliDateConverter.compareDates(
            from.toCore(), to.year.toInt(), to.month.toInt(), to.dayOfMonth.toInt()
        ).toLong()

    override fun getWeekdayName(
        dayOfWeek: Long,
        format: NameFormatDto,
        language: LangDto
    ): String = NepaliDateConverter.getWeekdayName(
        dayOfWeek.toInt(), format.toCore(), language.toCore()
    )

    override fun getBsMonthName(month: Long, format: NameFormatDto, language: LangDto): String =
        NepaliDateConverter.getMonthName(month.toInt(), format.toCore(), language.toCore())

    override fun getAdMonthName(month: Long, format: NameFormatDto, language: LangDto): String =
        NepaliDateConverter.getEnglishMonthName(month.toInt(), format.toCore(), language.toCore())

    override fun formatBsDate(calendar: CalendarDto, locale: LocaleDto): String =
        NepaliDateConverter.formatNepaliDate(calendar.toCore(), locale.toCore())

    override fun formatAdDate(calendar: CalendarDto, locale: LocaleDto): String =
        NepaliDateConverter.formatEnglishDate(calendar.toCore(), locale.toCore())

    override fun formatBsDateByPattern(pattern: String, date: DateDto, language: LangDto): String =
        NepaliDateConverter.formatNepaliDateByUnicodePattern(
            pattern,
            NepaliDateConverter.getNepaliCalendar(
                date.year.toInt(), date.month.toInt(), date.dayOfMonth.toInt()
            ),
            language.toCore()
        )

    override fun formatAdDateByPattern(pattern: String, date: DateDto, language: LangDto): String =
        NepaliDateConverter.formatEnglishDateByUnicodePattern(
            pattern,
            NepaliDateConverter.getEnglishCalendar(
                date.year.toInt(), date.month.toInt(), date.dayOfMonth.toInt()
            ),
            language.toCore()
        )

    override fun formatTimeByPattern(pattern: String, time: TimeDto, language: LangDto): String =
        NepaliDateConverter.formatTimeByUnicodePattern(pattern, time.toCore(), language.toCore())

    override fun formatBsDateTimeByPattern(
        pattern: String,
        date: DateDto,
        time: TimeDto?,
        language: LangDto
    ): String = NepaliDateConverter.formatNepaliDateTimeByUnicodePattern(
        pattern,
        NepaliDateConverter.getNepaliCalendar(
            date.year.toInt(), date.month.toInt(), date.dayOfMonth.toInt()
        ),
        time?.toCore(),
        language.toCore()
    )

    override fun formatAdDateTimeByPattern(
        pattern: String,
        date: DateDto,
        time: TimeDto?,
        language: LangDto
    ): String = NepaliDateConverter.formatEnglishDateTimeByUnicodePattern(
        pattern,
        NepaliDateConverter.getEnglishCalendar(
            date.year.toInt(), date.month.toInt(), date.dayOfMonth.toInt()
        ),
        time?.toCore(),
        language.toCore()
    )

    override fun formatTimeEnglish(time: TimeDto, use12HourFormat: Boolean): String =
        NepaliDateConverter.getFormattedTimeInEnglish(time.toCore(), use12HourFormat)

    override fun formatTimeNepali(time: TimeDto, use12HourFormat: Boolean): String =
        NepaliDateConverter.getFormattedTimeInNepali(time.toCore(), use12HourFormat)

    override fun bsDateTimeToIso(date: DateDto, time: TimeDto): String =
        NepaliDateConverter.formatNepaliDateTimeToIsoFormat(date.toCore(), time.toCore())

    override fun adDateTimeToIso(date: DateDto, time: TimeDto): String =
        NepaliDateConverter.formatEnglishDateNepaliTimeToIsoFormat(date.toCore(), time.toCore())

    override fun bsDateTimeFromIso(isoDateTime: String): DateTimeDto =
        NepaliDateConverter.getNepaliDateTimeFromIsoFormat(isoDateTime).toDto()

    override fun adDateTimeFromIso(isoDateTime: String): DateTimeDto =
        NepaliDateConverter.getEnglishDateNepaliTimeFromIsoFormat(isoDateTime).toDto()

    override fun localizeDigits(text: String, script: DigitScriptDto): String =
        with(NepaliDateConverter) { text.localizeDigits(script.toCore()) }

    override fun toLatinDigits(text: String): String =
        with(NepaliDateConverter) { text.toLatinDigits() }

    override fun wireFormatDate(
        date: DateDto,
        pattern: DatePatternDto,
        script: DigitScriptDto
    ): String = NepaliDateFormatter.format(date.toCore(), pattern.toCore(), script.toCore())

    override fun wireParseDate(input: String, pattern: DatePatternDto): DateDto? =
        NepaliDateFormatter.parse(input, pattern.toCore())?.let {
            DateDto(
                year = it.year.toLong(),
                month = it.month.toLong(),
                dayOfMonth = it.dayOfMonth.toLong()
            )
        }

    override fun wireFormatTime(time: TimeDto): String =
        NepaliTimeFormatter.format(time.toCore())

    override fun wireParseTime(input: String): TimeDto? =
        NepaliTimeFormatter.parse(input)?.toDto()

    override fun statusOf(policy: PolicyDto, date: DateDto): DayStatusDto =
        policy.toCore().statusOf(date.toCore()).toDto()

    override fun monthStatus(policy: PolicyDto, year: Long, month: Long): List<DayStatusDto> =
        policy.toCore().monthStatus(year.toInt(), month.toInt()).map { it.toDto() }

    override fun eventsOn(policy: PolicyDto, date: DateDto): List<EventDto> =
        policy.toCore().eventsOn(date.toCore()).map { it.toDto() }

    override fun eventsIn(policy: PolicyDto, year: Long, month: Long): List<EventDto> =
        policy.toCore().eventsIn(year.toInt(), month.toInt()).map { it.toDto() }

    override fun isNonWorkingDay(policy: PolicyDto, date: DateDto): Boolean =
        policy.toCore().isNonWorkingDay(date.toCore())

    override fun workingDaysBetween(policy: PolicyDto, start: DateDto, end: DateDto): Long =
        NepaliDateConverter.workingDaysBetween(
            start.toCore(), end.toCore(), policy.toCore()
        ).toLong()

    override fun nextWorkingDay(policy: PolicyDto, from: DateDto): CalendarDto =
        NepaliDateConverter.nextWorkingDay(from.toCore(), policy.toCore())
            .let { NepaliDateConverter.getNepaliCalendar(it.year, it.month, it.dayOfMonth) }
            .toDto()

    override fun addWorkingDays(policy: PolicyDto, from: DateDto, days: Long): CalendarDto =
        NepaliDateConverter.addWorkingDays(from.toCore(), days.toInt(), policy.toCore())
            .let { NepaliDateConverter.getNepaliCalendar(it.year, it.month, it.dayOfMonth) }
            .toDto()

    override fun eventSpanningDays(event: EventDto, days: Long): List<EventDto> =
        event.toCore().spanningDays(days.toInt()).map { it.toDto() }

    override fun eventSpanningThrough(event: EventDto, end: DateDto): List<EventDto> =
        event.toCore().spanningThrough(end.toCore()).map { it.toDto() }
}
